arangod/Cluster/CreateCollection.cpp

```diff
@@ -0,0 +1,155 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2018 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Kaveh Vahedipour
+/// @author Matthew Von-Maszewski
+////////////////////////////////////////////////////////////////////////////////
+
+#include "CreateCollection.h"
+#include "MaintenanceFeature.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "Basics/VelocyPackHelper.h"
+#include "Cluster/ClusterFeature.h"
+#include "Cluster/FollowerInfo.h"
+#include "VocBase/LogicalCollection.h"
+#include "VocBase/Methods/Collections.h"
+#include "VocBase/Methods/Databases.h"
+
+#include <velocypack/Compare.h>
+#include <velocypack/Iterator.h>
+#include <velocypack/Slice.h>
+#include <velocypack/velocypack-aliases.h>
+
+
+using namespace arangodb::application_features;
+using namespace arangodb::maintenance;
+using namespace arangodb::methods;
+
+constexpr auto WAIT_FOR_SYNC_REPL = "waitForSyncReplication";
+constexpr auto ENF_REPL_FACT = "enforceReplicationFactor";
+
+CreateCollection::CreateCollection(
+  MaintenanceFeature& feature, ActionDescription const& desc)
+  : ActionBase(feature, desc) {
+
+  //Todo: runtime behaviour
+  
+  TRI_ASSERT(desc.has(SHARD));
+  TRI_ASSERT(desc.has(COLLECTION));
+  TRI_ASSERT(desc.has(DATABASE));
+  TRI_ASSERT(desc.has(LEADER));
+  TRI_ASSERT(properties().hasKey(TYPE));
+  TRI_ASSERT(properties().get(TYPE).isInteger());  
```

Comment:
How about moving these TRI_ASSERTs() to a validateInput() function (a virtual function on ActionBase).  And have ActionBase::create() call it on every new object.  The fail the object there instead of hoping we caught mistakes by compiling in MAINTAINER mode.

Reply:
Or set an isValid bool to true after testing all these items in the constructor.  Let create() then test if new action isValid.